messaging/coap/observe.c

```diff
@@ -888,194 +887,275 @@ notify_resource_defaults_observer(oc_resource_t *resource,
+    oc_discovery_create_batch_for_resource(&links_array, batch_obs->resource,
+                                           &obs->endpoint);
+    while (o != NULL) {
+      batch_observer_t *next = o->next;
+      if (o->obs == obs) {
+        oc_discovery_create_batch_for_resource(&links_array, o->resource,
+                                               &obs->endpoint);
+        oc_list_remove(batch_observers_list, o);
+        oc_memb_free(&batch_observers_memb, o);
+      }
+      o = next;
+    }
+    int size_after = oc_rep_get_encoded_payload_size();
+    if (size_before == size_after) {
+      OC_DBG("process_batch_observers: acls(%d) drop observations\n\n");
+    } else {
+      oc_rep_end_links_array();
+      size_after = oc_rep_get_encoded_payload_size();
+      if (size_after < 0) {
+        OC_ERR("process_batch_observers: invalid size after batch "
+               "serialization\n\n");
+      } else {
+        OC_DBG("process_batch_observers: sending data with size %d\n\n",
+               size_after);
+        response_buffer.content_format = APPLICATION_VND_OCF_CBOR;
+        response_buffer.response_length = size_after;
+        response_buffer.code = oc_status_code(OC_STATUS_OK);
+        oc_response_t response = { 0 };
+        response.response_buffer = &response_buffer;
+        if (send_notification(obs, &response, obs->resource->uri, true,
+                              dispatch_process_batch_observers)) {
+          goto leave_notify_observers;
         }
-        coap_set_token(notification, obs->token, obs->token_len);
-        transaction = coap_new_transaction(coap_get_mid(), obs->token,
-                                           obs->token_len, &obs->endpoint);
-        if (transaction) {
-          obs->last_mid = transaction->mid;
-          notification->mid = transaction->mid;
-          transaction->message->length =
-            coap_serialize_message(notification, transaction->message->data);
-          if (transaction->message->length > 0) {
-            coap_send_transaction(transaction);
-          } else {
-            coap_clear_transaction(transaction);
-          }
-        } // transaction
-      }   // response_buf != NULL
-    }     //! separate response
-    obs = obs->next;
-  } // While observers
+      }
+    }
+    oc_list_remove(batch_observers_list, batch_obs);
+    oc_memb_free(&batch_observers_memb, batch_obs);
+    batch_obs = (batch_observer_t *)oc_list_head(batch_observers_list);
+  }
 leave_notify_observers:;
 #ifdef OC_DYNAMIC_ALLOCATION
   if (buffer) {
     free(buffer);
   }
-#endif
+#endif /* OC_DYNAMIC_ALLOCATION */
+  return OC_EVENT_DONE;
+}
+
+void
+coap_notify_discovery_batch_observers(oc_resource_t *resource)
+{
+  if (resource == NULL) {
+    return;
+  }
+  oc_resource_t *discover_resource =
+    oc_core_get_resource_by_index(OCF_RES, resource->device);
+
+  if (discover_resource == resource) {
+    return;
+  }
+
+  coap_observer_t *obs = NULL;
```

Comment:
Can be moved inside the for-loop, it's not used outside of it.

Reply:
moved